### DISCUSSION TITLE
Parent of @Serializable, how to obtain it?

### DISCUSSION THREAD

**Discussion by chumpa** [pdvrieze/xmlutil#64] [Q&A]
Suppose I have two classes:

```kotlin
@Serializable
class A(
    @XmlElement(true)
    val b: B,
)

@Serializable
class B(
    @XmlElement(true)
    val contentB: String,
) {
    var a:A? = null
    init {
// How can I obtain reference to parent A here?
        println("contentB=$contentB, A=$a")
    }
}

val a = xml.decodeFromString<A>("<A><B><contentB>content</contentB></B></A>")

```

I need some way to get instance of A in init block of B when decoding legacy XMLs. Is there any way?

**Comment by pdvrieze** [comment]
Unfortunately this is not possible directly. The way that the framework/library works is that it will first create all members (`B`), collect them and then call the constructor of the container (`A`). In other words, you can't as it doesn't exist.

This is a place where you will probably want to use some delegate for serialization rather than use a production class directly (Note that this delegate type could be file private or internal). So you basically have a very dumb data type that you (de)serialize (for A and/or B) and then you use that type to create the final structure (with reference to container/parent).

You can make this transparent by creating a custom serializer for `A` that deserializes `A` by deserializing the delegate (and maybe a delegate for B) and then calls the constructor for A with this delegate as parameter. The constructor for A can then call the constructor B with itself as one of the parameters.

**Reply by EMaksymenko** [reply]
@pdvrieze Is is possible to post-process de-serialized class somehow and assign transient parent attribute when we already have parent object de-serialized?
I have a similar problem as topic author, but I do not need to have parent in init block. I need to address parent of class later in business logic.

**Reply by pdvrieze** [reply]
Yes it is possible to postprocess. The way I would do this is to have a `ASerialDelegate` private member class, and then have a custom deserializer for `A` that deserializes through the delegate, and then postprocesses that to create an `A` instance.
